res/controllers/MVave-SMC-Mixer-scripts.js

```diff
@@ -0,0 +1,469 @@
+                midi: [0x90, 0x5B],
+                key: "beatjump_backward",
+            });
+            this.forwardButton = new components.Button({
+                group: "[Channel1]",
+                midi: [0x90, 0x5C],
+                key: "beatjump_forward",
+            });
+        }
+    }
+
+    class Encoder extends components.Encoder {
+        constructor(params) {
+            super(params);
+            this.input = function(_channel, _control, value, _status, _group) {
+                if (value === 0x41) {
+                    this.inSetParameter(this.inGetParameter() - 0.01);
+                } else if (value === 0x01) {
+                    this.inSetParameter(this.inGetParameter() + 0.01);
+                }
+            };
+        }
+    }
+    // LongPressButton is like a normal button of type powerWindow, except that it
+    // doesn't trigger the short press and the long press. Instead it triggers on
+    // release and leaves it to the user of the class to check if this was a long
+    // press or a short press.
+    class LongPressButton extends components.Button {
+        constructor(knob, params) {
+            super(params);
+            this.input = function(channel, control, value, status, _group) {
+                if (this.isPress(channel, control, value, status)) {
+                    this.isLongPressed = false;
+                    this.longPressTimer = engine.beginTimer(this.longPressTimeout, () => {
+                        this.isLongPressed = true;
+                        this.longPressTimer = 0;
+                    }, true);
+                } else {
+                    this.inToggle();
+                    if (!this.isLongPressed && this.triggerOnRelease) {
+                        this.trigger();
+                    }
+                    if (this.longPressTimer !== 0) {
+                        engine.stopTimer(this.longPressTimer);
+                        this.longPressTimer = 0;
+                    }
+                    this.isLongPressed = false;
+                }
+            };
+            this.inToggle = function() {
+                if (this.isLongPressed) {
+                    if (this.knob.key === this.key.replace("button_", "") || (this.key === "enabled" && this.knob.key === "super1")) {
+                        this.knob.group = this.origGroup;
+                        this.knob.key = this.origKey;
+                        this.knob.inKey = this.origInKey;
+                        this.knob.outKey = this.origOutKey;
+                    } else {
+                        this.knob.group = this.group;
+                        let newKey = "";
+                        if (this.key === "enabled") {
+                            newKey = "super1";
+                        } else {
+                            newKey = this.key.replace("button_", "");
+                        }
+                        this.knob.key = newKey;
+                        this.knob.inKey = newKey;
```

Comment:
can you explain some of the special sauce you do here? Why is it necessary? What are you trying to accomplish?

Reply:
I'll add a comment to make it clearer. The general gist is that there's only one rotary encoder per knob at the top, and we want to be able to use it for all the EQ buttons, but also have cutoffs. Short pressing the button does the cutoff, and long pressing changes the knob to be that EQ control. If we use the regular power window the short press always fires, so the cutoff is turned on and you hear it even though that's not what you meant to do.

Reply:
Oh, or you may have meant all the "button" and "super1" stuff. That's to set the knob to change the value affected by the button (ie. if we hit `button_parameter1` we should have the knob's key set to `parameter1`, or for `enabled` we want it set to `super1`, the quick effects button). I agree this is confusing. I'll see if I can't think of a way to simplify it, maybe just with a big switch statement that would be easier to read, or add a comment if not.

Reply:
So the pot changes behavior depending on the button state? Then there are two established ways to deal with that: 
1. Instance two separate components and then route the `input` where you want it to go. You can do that by simply having both as local variable and then assign the instance you want to dispatch to in the button handler. 
2. You look at how the `shift` and `unshift` stuff is done. Each component has a function that reconfigures it on the fly. Then you would do a `c.disconnect(); c.shift(); c.connect(); c.trigger();` (where `c` is the component instance you need to reconfigure). In your case, you would set new `inKey`s and `group`s. 

Does that make sense?

Reply:
Its primary a codestyle question. The class defines the behavior and the instance supplies the data. In your current solution, the both are being conflated because the class makes explicit assumptions on what values (COs) its being used with. So I don't consider the current solution particular clean.

Reply:
Oh right, I see what you're saying; I think I had it the "correct" way originally, then got lazy and decided the class was only used in this one file anyways so I changed it so that I wouldn't have to assign the method to every instance.

I'm generally a fan of not pre-generalizing something that will never be used in any other file just for the sake of technically being more robust to future use, but I'll go try to swap it around and see if I can make it work in a clean way.

Reply:
yeah, definitely good guidance, but in this case the generalization would result in stricter separation of concerns and thus increased code clarity. I wouldn't worry too much about duplication of the instances right now, its very easy to combat that (use loops and LUTs, or merge partial Object definitions using `Object.assign`).

Reply:
I moved this back to where each instance is defined and curried in the various values it needs to store so that I only had to define the function in one place. Hopefully that's a little cleaner; LMK what you think, and thanks for all your help with this review!